observatory/tracking.py

```diff
@@ -1,51 +1,247 @@
         """
-        self.url = url
+        
+        if settings is None:
+            warnings.warn(
+                'Trying to record empty settings. ' + 
+                'To prevent risk of settings loss on the server, ' +
+                'the empty settings collection is discarded.', 
+                RuntimeWarning)
 
+        self._state.record_settings(
+            self.name, self.version, self.experiment,
+            self.run_id, dict(settings))
+
+    def record_output(self, input_file, filename):
+        """
+        Records an output for the current run.
+
+        Parameters
+        ----------
+        input_file : object
+            Filename or handle to input file
+        filename : string
+            Name of the file as it should be stored on the server
+        """
+
+        if filename is None or filename.strip() == '':
+            raise AssertionError(
+                'Please provide a valid filename to store the output on the server.')
+
+        if input_file is None or input_file.strip() == '':
+            raise AssertionError(
+                'Please provide a valid filename for input_file.')
+
+        absolute_file_path = path.abspath(input_file)
+
+        if not path.exists(absolute_file_path):
+            raise AssertionError(
+                f'Could not find source file {absolute_file_path} ' +
+                'to upload as output of this run. Please make ' +
+                'sure that the file exists on disk.')
+
+        self._state.record_output(
+            self.name, self.version, self.experiment,
+            self.run_id, filename, open(absolute_file_path))
+
+    def __enter__(self):
+        self._state.record_session_start(
+            self.name, self.version, self.experiment, self.run_id)
+
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if exc_type is None:
+            session_status = 'COMPLETED'
+        else:
+            session_status = 'FAILED'
+
+        self._state.record_session_end(
+            self.name, self.version, self.experiment,
+            self.run_id, session_status)
+
+
+        return exc_type is None
+
+class ObservatoryState(ABC):
+    def __init__(self):
+        self.n = None
+
+    def switch(self, state):
+        self.__class__ = state
+
+    @abstractmethod
+    def record_metric(self, name, value):
+        """
+        This method will record a metric.
+        """
+        pass
+
+    @abstractmethod
+    def record_settings(self, name, value):
+        """
+        This method will record a setting.
+        """
+        pass
+
+    @abstractmethod
+    def record_output(self, name, value):
+        """
+        This method will record a output.
+        """
+        pass
+
+    @abstractmethod
     def record_session_start(self, model, version, experiment, run_id):
         """
-        Records the start of a session.
+        This method will record the start of a run.
```

Comment:
```suggestion
Override this method in a derived class to record the start of a session.
The derived class is required to set the status of the run to "In progress." with a unix timestamp.
```

Please include a description of the parameters as well.